### DISCUSSION TITLE
Use a `.cake` folder by default instead of a `tools` folder (?)

### DISCUSSION THREAD

**Discussion by augustoproiete** [cake-build/cake#2979] [Ideas]
#### Idea/suggestion for a future version of Cake:

As we all know, Cake uses a folder for downloading and extracting NuGet packages (and other things), and by default it uses a `./tools` folder.

There a couple of (small) issues I've seen because of that:

1. **Adding Cake to existing projects** (any language/platform)
    `./tools` is a [common enough name that some projects use for other purposes](https://github.com/cake-build/website/issues/468) (even internal to the project) and when introducing Cake to existing projects that already define a `./tools` folder, one needs to be careful to ensure Cake won't touch that folder and will use a different folder instead, which is not as straightforward as changing the `cake.config` as sometimes that means [adapting bootstrappers that don't honor what's set in the `cake.config`](https://github.com/cake-build/resources/issues/84).

2. **Adding Cake to new projects** (.NET-specific)
    When starting a new project, it's common for devs to use the default [`VisualStudio.gitignore`](https://github.com/github/gitignore/blob/6eff882467cbde7b4e293b4dd6abd7685fc67844/VisualStudio.gitignore) on GitHub. The good news is that [this .gitignore already brings a couple of lines specific to Cake](https://github.com/github/gitignore/blob/6eff882467cbde7b4e293b4dd6abd7685fc67844/VisualStudio.gitignore#L318-L320). The bad news is that [they are commented out](https://github.com/github/gitignore/blob/6eff882467cbde7b4e293b4dd6abd7685fc67844/VisualStudio.gitignore#L318-L320), because having a `./tools` by default would be too broad of a rule.

If Cake started using a `.cake` folder (or similar), it would avoid the issues above as it's unlikely users would have a `.cake` folder for a different purpose other than using Cake, and it would be specific enough to be included in the [`VisualStudio.gitignore`](https://github.com/github/gitignore/blob/6eff882467cbde7b4e293b4dd6abd7685fc67844/VisualStudio.gitignore) template.

Thoughts?

**Comment by pascalberger** [comment]
We have an open issue for using global packages folder (https://github.com/cake-build/cake/issues/2830) which I think would solve the mentioned points

**Reply by augustoproiete** [reply]
Does that completely eliminate the need for Cake having a "work" folder that is local to the project?

**Reply by pascalberger** [reply]
There are a few difficult things to solve like tool and module resolving. If this could be solved all NuGet packages are out. Leaving only bootstrapper specific files (`packages.config`, `nuget.exe`) in there which are not strictly required.

**Reply by augustoproiete** [reply]
Given this seems to be something for the v2.0.0 timeframe, wouldn't it be worth (as an intermediary step) to:

- Update the [cake.config template](https://github.com/cake-build/resources/blob/327d2f71495071537a21529f31edef66017daf1b/cake.config#L10-L12) to use `./.cake` instead `./tools`
- Update the [.NET Framework bootstrapper templates](https://github.com/cake-build/resources/blob/2ee42aa524adf0a3679e5e136a640fb1df06ab15/dotnet-framework/build.ps1#L118) to use `./.cake` instead `./tools`

?

**Reply by pascalberger** [reply]
We don't want to make this change in the 1.x timeframe since it probably will break more people than it benefits. It's also not done with updating just these two files, since the new defaults would need to be propagated through the whole ecosystem.

2.0 will be the release where we can think about making changes like this (and we don't expect 2.0 to take another 6 years 😉)

**Comment by RehanSaeed** [comment]
I came here to make this exact suggestion. Props to @augustoproiete, you got here first!

**Comment by augustoproiete** [comment]
Tracking via https://github.com/cake-build/cake/issues/3263 scheduled for ~Cake v2.0~ Cake v3.0
